apiextensions/v1alpha1/types_compatibilityrequirement.go

```diff
@@ -0,0 +1,391 @@
+// It is used to ensure compatibility between different actors using the same
+// CRD.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +openshift:file-pattern=cvoRunLevel=0000_20,operatorName=crd-compatibility-checker,operatorOrdering=01
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=compatibilityrequirements,scope=Cluster
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/2479
+type CompatibilityRequirement struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitzero"`
+
+	// spec is the specification of the desired behavior of the Compatibility Requirement.
+	// +required
+	Spec CompatibilityRequirementSpec `json:"spec,omitzero"`
+
+	// status is the most recently observed status of the Compatibility Requirement.
+	// +optional
+	Status CompatibilityRequirementStatus `json:"status,omitzero"`
+}
+
+// CompatibilityRequirementSpec is the specification of the desired behavior of the Compatibility Requirement.
+type CompatibilityRequirementSpec struct {
+	// compatibilitySchema defines the schema used by
+	// customResourceDefinitionSchemaValidation and objectSchemaValidation.
+	// This field is required.
+	// +required
+	CompatibilitySchema CompatibilitySchema `json:"compatibilitySchema,omitempty,omitzero"`
+
+	// customResourceDefinitionSchemaValidation ensures that updates to the
+	// installed CRD are compatible with this compatibility requirement. If not
+	// specified, admission of the target CRD will not be validated.
+	// This field is optional.
+	// +optional
+	CustomResourceDefinitionSchemaValidation CustomResourceDefinitionSchemaValidation `json:"customResourceDefinitionSchemaValidation,omitempty,omitzero"`
+
+	// objectSchemaValidation ensures that matching resources conform to
+	// compatibilitySchema. If not specified, admission of matching resources
+	// will not be validated.
+	// This field is optional.
+	// +optional
+	ObjectSchemaValidation ObjectSchemaValidation `json:"objectSchemaValidation,omitempty,omitzero"`
+}
+
+// CRDDataType indicates the type of the CRD data.
+type CRDDataType string
+
+const (
+	// CRDDataTypeYAML indicates that the CRD data is in YAML format.
+	CRDDataTypeYAML CRDDataType = "YAML"
+)
+
+// CRDData contains the complete definition of a CRD.
+type CRDData struct {
+	// type indicates the type of the CRD data. The only supported type is YAML.
+	// This field is required.
+	// +kubebuilder:validation:Enum=YAML
+	// +required
+	Type CRDDataType `json:"type,omitempty"`
```

Comment:
In the future, what are the odds that this field might need to have a sibling that is not data, e.g. from a reference or similar? Just wondering if we want to make this a union type rather than type and data 🤔 

```
customResourceDefintion:
  type: yaml | json | ConfigMap
  yaml:
    data: ...
  json:
    data: ...
  configMap:
    name: ...
    key: ...
```